### PR TITLE
Build system: support alternate LIBDIR, MANDIR, DOCDIR

### PR DESCRIPTION
**Pull request by rworkman**
This addresses #926 and #927

### CODE REVIEW

File: programs/Makefile [hunk 1]
```diff
@@ -262,9 +262,7 @@ mandir      ?= $(datarootdir)/man
 man1dir     ?= $(mandir)/man1
 
 ifneq (,$(filter $(shell uname),OpenBSD FreeBSD NetBSD DragonFly SunOS))
-MANDIR  ?= $(PREFIX)/man/man1
-else
-MANDIR  ?= $(man1dir)
+MANDIR  ?= $(PREFIX)/man
```
Comment (Cyan4973) [inline programs/Makefile]: This construction prevents using `mandir` nor `man1dir` as an input of this Makefile.
These variables are recommended by the [GNU Makefile coding standards](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html#Directory-Variables).

It also changes the destination directory for non-BSD targets, 
from `$(prefix)/share/man/man1` to `$(prefix)/man/man1`.

File: programs/Makefile [hunk 2]
```diff
@@ -281,17 +279,17 @@ INSTALL_MAN     ?= $(INSTALL_DATA)
 .PHONY: install
 install: zstd
 	@echo Installing binaries
-	@$(INSTALL) -d -m 755 $(DESTDIR)$(BINDIR)/ $(DESTDIR)$(MANDIR)/
+	@$(INSTALL) -d -m 755 $(DESTDIR)$(BINDIR)/ $(DESTDIR)$(MANDIR)/man1
```
Comment (Cyan4973) [inline programs/Makefile]: Is there an issue inserting `/man1` directly into `$MANDIR` ?
Would you prefer such variable to be called `$MAN1DIR` instead ?

File: build/cmake/programs/CMakeLists.txt [hunk 3]
```diff
@@ -45,9 +45,9 @@ IF (UNIX)
         COMMENT "Copying manpage zstd.1")
     ADD_CUSTOM_TARGET(zstdcat.1 ALL ${CMAKE_COMMAND} -E create_symlink zstd.1 zstdcat.1 DEPENDS zstd.1 COMMENT "Creating zstdcat.1 symlink")
     ADD_CUSTOM_TARGET(unzstd.1 ALL ${CMAKE_COMMAND} -E create_symlink zstd.1 unzstd.1 DEPENDS zstd.1 COMMENT "Creating unzstd.1 symlink")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstd.1 DESTINATION "share/man/man1")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstdcat.1 DESTINATION "share/man/man1")
-    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/unzstd.1 DESTINATION "share/man/man1")
+    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstd.1 DESTINATION "${CMAKE_INSTALL_MANDIR}/man1")
```
Comment (Cyan4973) [inline build/cmake/programs/CMakeLists.txt]: why removing `share` from the man install path ?

File: Makefile [hunk 4]
```diff
@@ -14,6 +14,11 @@ ZWRAPDIR = zlibWrapper
 TESTDIR  = tests
 FUZZDIR  = $(TESTDIR)/fuzz
 
+# Define some install locations
```
Comment (Cyan4973) [inline Makefile]: Why defining `cmake` variables within the root `Makefile` ?

Comment (Cyan4973) [inline Makefile]: If these variables are only used by `cmake`,
maybe it's better for code maintenance to define them close to the place where they are used.

File: Makefile [hunk 5]
```diff
@@ -14,6 +14,11 @@ ZWRAPDIR = zlibWrapper
...
+LIBDIR   = $(CMAKE_INSTALL_PREFIX)/lib
+MANDIR   = $(CMAKE_INSTALL_PREFIX)/share/man
```
Comment (Cyan4973) [inline Makefile]: My understanding is that `MANDIR` is different depending on Linux / BSD targets.
Here it seems there is no distinction.

File: build/cmake/programs/CMakeLists.txt [hunk 6]
```diff
@@ -45,9 +45,9 @@ IF (UNIX)
...
+    INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/zstdcat.1 DESTINATION "${CMAKE_INSTALL_MANDIR}/man/man1")
```
Comment (Cyan4973) [inline build/cmake/programs/CMakeLists.txt]: I guess you meant `"${CMAKE_INSTALL_MANDIR}/man1"` ?

**Comment by Cyan4973** [comment]
How do you test the result of your installation script ?

File: programs/Makefile (hunk 1 above)
Comment (rworkman) [reply programs/Makefile]: Yeah, this was unintentional.

File: programs/Makefile (hunk 2 above)
Comment (rworkman) [reply programs/Makefile]: Disclaimer: this is my first time messing with a cmake-ish (but it's really not even that) build system; from what I recall from messing with autotools and even much older Makefiles, I've never seen MANDIR have the man? subdir in it -- it's always been $(MANDIR)/man?/

File: build/cmake/programs/CMakeLists.txt (hunk 6 above)
Comment (rworkman) [reply build/cmake/programs/CMakeLists.txt]: Indeed.

**Comment by rworkman** [comment]
The build process here is essentially this:

    patch -p1 < $CWD/0001-Build-system-support-alternate-LIBDIR-MANDIR-DOCDIR.patch
    
    make \
      PREFIX=/usr \
      LIBDIR=/usr/lib${LIBDIRSUFFIX} \
      MANDIR=/usr/man \
      DOCDIR=/usr/doc/$PKGNAM-$VERSION
    
    make install \
      PREFIX=/usr \
      LIBDIR=/usr/lib${LIBDIRSUFFIX} \
      MANDIR=/usr/man \
      DOCDIR=/usr/doc/$PKGNAM-$VERSION \
      DESTDIR=$PKG

About to push a new commit. I'm not convinced that it addresses the comment about where the variables are defined, but maybe it's acceptable now since they aren't specific to cmake any more.

**Comment by rworkman** [comment]
This iteration drops DOCDIR (it wasn't used - I had planned to use it, but meh - too much trouble), add PREFIX, and addresses the review points, I think.

**Comment by Cyan4973** [comment]
I'm not comfortable applying this patch.
I do believe it does improve situation on your system,
but its impact seems too wide and affect different systems in hard to guess ways.

Please note that creating and maintaining a build/install script for a large number of systems is a fairly difficult exercise. While nothing is perfect, it can certainly be improved. When visibility is poor, it's important to progress cautiously by tiny steps, with limited / manageable scope.

I would be ok to start with a patch impacting only `cmake` to solve #926.

**Pull request closed without merging**